compiler/src/dmd/toir.d

```diff
@@ -967,6 +969,181 @@ void buildClosure(FuncDeclaration fd, IRState *irs)
+    {
+        if (v.inClosure)
+            continue;
+
+        /* Align and allocate space for v in the align closure
+         * just like AggregateDeclaration.addField() does.
+         */
+        const memsize = cast(uint)v.type.size();
+        const memalignsize = v.type.alignsize();
+        const xalign = v.alignment;
+
+        AggregateDeclaration.alignmember(xalign, memalignsize, &offset);
+        v.offset = offset;
+        //printf("align closure var %s, offset = %d\n", v.toChars(), offset);
+
+        offset += memsize;
+
+        uint actualAlignment = xalign.isDefault() ? memalignsize : xalign.get();
+        //printf("actualAlignment = x%x, x%x\n", actualAlignment, xalign.get());
+        if (aggAlignment < actualAlignment)
+            aggAlignment = actualAlignment;     // take the largest
+    }
+
+    return aggAlignment;
+}
+
+/*************************************
+ * Aligned sections are implemented by taking the local variables that
+ * need alignment that is larger than the stack alignment.
+ * They are allocated into a separate chunk of memory on the stack
+ * called an align section, which is aligned on function entry.
+ *
+ * buildAlignSection() inserts code just after the function prolog
+ * is complete. It allocates memory for the align closure by making
+ * a local stack variable to contain that memory, allocates
+ * a local variable (salignSection) to point to it.
+ * In VarExp::toElem and SymOffExp::toElem, when referring to a
+ * variable that is in an align closure, take the offset from salignSection rather
+ * than from the frame pointer.
+ * A variable cannot be in both a closure and an align section. They go in the closure
+ * and then that closure is aligned.
+ *
+ * getEthis() and NewExp::toElem need to use sclosure, if set, rather
+ * than the current frame pointer??
+ *
+ * Run after buildClosure, as buildClosure gets first dibs on inAlignSection variables
+ * Params:
+ *      fd = function in which all this occurs
+ *      irs = state of the intermediate code generation
+ * Reference:
+ *      buildClosure() is very similar.
+ *
+ *      https://github.com/dlang/dmd/pull/9143 was an incomplete attempt to solve this problem
+ *      that was merged. It should probably be removed.
+ */
+void buildAlignSection(FuncDeclaration fd, ref IRState irs)
+{
+    enum log = false;
+    if (log) printf("buildAlignSection(fd = %s)\n", fd.toChars());
+    if (!fd.alignSectionVars)
+        return;
+    auto alignSectionVars = (*fd.alignSectionVars)[];
+
+    /* If they're all in a closure, don't need to build an align section
+     */
+    foreach (v; alignSectionVars)
+    {
+        if (!v.inClosure)
+            goto L1;
+    }
+    return;
+  L1:
+
+    auto stackAlign = target.stackAlign();
+    auto aggAlignment = setAlignSectionVarOffset(fd);
+
+    /* Generate type name for align closure struct */
+    const char *name1 = "ALIGNSECTION.";
+    const char *name2 = fd.toPrettyChars();
```

Comment:
Is this supposed to be unique? Because overloads can have the same 'pretty' name.

Reply:
It doesn't need to be unique. This way it is easily readable in the debug info.